File: Hash/202.快乐数.cpp
/*
 * @lc app=leetcode.cn id=202 lang=cpp
 *
 * [202] 快乐数
 */

// @lc code=start
class Solution
{
public:
    // 求和
    int getSum(int n)
    {
        int sum = 0;

        while (n > 0)
        {
            int singleNum = n % 10;
            n = n / 10;
            sum += singleNum * singleNum;
        }

        return sum;
    }

    bool isHappy(int n)
    {
        unordered_set<int> result;

        int sum = getSum(n);
        // 和 != 1，表示需要继续往下计算
        while (sum != 1)
        {
            // 和已经存在，说明之前已经生成过
            // 一段循环之后继续生成这个结果，永远不可能是快乐数
            if (result.find(sum) != result.end())
            {
                return false;
            }
            // 求和结果追加到set
            result.insert(sum);
            // 计算新结果
            sum = getSum(sum);
        }

        return true;
    }
};
// @lc code=end

File: Hash/202.快乐数.ts
/*
 * @lc app=leetcode.cn id=202 lang=typescript
 *
 * [202] 快乐数
 */

// @lc code=start
function isHappy(n: number): boolean {
  // sum是下一次拆分进行平方的原始数据
  // 所以这里使用Object记录sum是否出现过，出现贵的话就会陷入死循环，一定不是快乐树
  let result = {}
  let sum = 0

  // 存在且不止出现过一次，死循环了，肯定不是快乐数，return false
  while (!result[n] || result[n] === 1) {
    // 求和开始
    while (n > 9) {
      let temp = n % 10
      n = Math.floor(n / 10)
      sum += Math.pow(temp, 2)
    }
    sum += Math.pow(n, 2)
    // 等于1，退出循环
    if (sum === 1) {
      return true
    }

    // 已经出现过一次，变成第二次
    if (result[sum] === 1) {
      result[sum] = 2
    } else {
      // 没有出现过，记录第一次
      result[sum] = 1
    }

    n = sum
    sum = 0
  }

  return false
};
// @lc code=end

